### PR TITLE
Rework bots to use JSON API

### PR DESCRIPTION
**Pull request by tamaskalcza-da**

### CODE REVIEW

File: Dockerfile-json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2019, Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+#
+
+ARG sdk_vsn=0.13.52
```
Comment (richardkapolnai-da) [inline Dockerfile-json]: include maintenance upgrade script

File: Dockerfile-json [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+FROM digitalasset/daml-sdk:${sdk_vsn}
+
+WORKDIR /home/daml
+
+USER daml
+
+EXPOSE 7575
+
+ENTRYPOINT ~/.daml/bin/daml json-api --ledger-host ${SANDBOX_HOST} --ledger-port ${SANDBOX_PORT} --http-port 7575 --max-inbound-message-size 4194304 --package-reload-interval 5s --application-id HTTP-JSON-API-Gateway
```
Comment (richardkapolnai-da) [inline Dockerfile-json]: optional: omit defaults if possible

File: pom.xml [hunk 3]
```diff
@@ -59,6 +71,33 @@
...
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-api</artifactId>
+            <version>0.10.7</version>
+        </dependency>
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-impl</artifactId>
+            <version>0.10.7</version>
+            <scope>runtime</scope>
```
Comment (richardkapolnai-da) [inline pom.xml]: optional: clarify/verify

File: src/main/java/com/digitalasset/jsonapi/Utils.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import java.time.Duration;
+import java.time.Instant;
+import org.apache.http.client.fluent.Request;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Utils {
+  private static final Logger log = LoggerFactory.getLogger("JsonAPI Util");
+
+  public static void waitForJsonApi(URI uri) throws Exception {
```
Comment (richardkapolnai-da) [inline src/main/java/com/digitalasset/jsonapi/Utils.java]: optional: refactor with eventually

File: src/main/java/com/digitalasset/jsonapi/events/CreatedEvent.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+  public Identifier getTemplateId() {
+    return templateId;
+  }
+
+  public String getContractId() {
+    return contractId;
+  }
+
+  public Template getPayload() {
```
Comment (richardkapolnai-da) [inline src/main/java/com/digitalasset/jsonapi/events/CreatedEvent.java]: optional: remove?

File: src/main/java/com/digitalasset/jsonapi/gson/GsonDeserializer.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import com.digitalasset.jsonapi.http.HttpResponse;
+import com.digitalasset.jsonapi.http.WebSocketResponse;
+import com.digitalasset.jsonapi.json.JsonDeserializer;
+import com.google.gson.Gson;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+public class GsonDeserializer {
+
+  private final Gson gson = GsonRegisteredAllDeserializers.gson();
```
Comment (richardkapolnai-da) [inline src/main/java/com/digitalasset/jsonapi/gson/GsonDeserializer.java]: inline if works?

File: src/main/java/com/digitalasset/jsonapi/gson/ResultDeserializer.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+  @Override
+  public Result deserialize(
+      JsonElement jsonElement, Type type, JsonDeserializationContext jsonDeserializationContext)
+      throws JsonParseException {
+    return jsonDeserializationContext.deserialize(jsonElement, dispatch(jsonElement));
+  }
+
+  private Type dispatch(JsonElement jsonElement) {
+    if (jsonElement.isJsonObject()) {
```
Comment (richardkapolnai-da) [inline src/main/java/com/digitalasset/jsonapi/gson/ResultDeserializer.java]: optional: rewrite?

File: src/main/java/com/digitalasset/jsonapi/tyrus/TyrusWebSocketClient.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  private Flowable<WebSocketResponse> createWebSocketPublisher(URI resource, String query)
+      throws IOException, DeploymentException {
+    PublishProcessor<String> broadcaster = PublishProcessor.create();
+    Session session =
+        client.connectToServer(new EmittingWebSocketEndpoint(broadcaster), config, resource);
+    session.getBasicRemote().sendText(query, true);
+    return broadcaster.map(this::toWebSocketResponse);
+  }
+
+  private WebSocketResponse toWebSocketResponse(String message) {
```
Comment (richardkapolnai-da) [inline src/main/java/com/digitalasset/jsonapi/tyrus/TyrusWebSocketClient.java]: s/string/inputstream/ ?

File: src/main/java/com/digitalasset/refapps/marketdataservice/Main.java [hunk 9]
```diff
@@ -4,107 +4,109 @@
...
-        scheduler = Executors.newScheduledThreadPool(1);
-        timeUpdaterBotExecutor = new TimeUpdaterBotExecutor(scheduler);
-        timeUpdaterBotExecutor.start(timeUpdaterBot, systemPeriodTime);
-      }
-    };
+  private static void wire(
+      AppConfig appConfig,
+      String party,
+      ContractQuery contractQuery,
+      // TODO: Don't we need a Flowable<Collection<Command>>?
```
Comment (richardkapolnai-da) [inline src/main/java/com/digitalasset/refapps/marketdataservice/Main.java]: remove TODO?

File: src/main/java/com/digitalasset/refapps/marketdataservice/timeservice/TimeUpdaterBot.java [hunk 10]
```diff
@@ -63,66 +49,32 @@ void updateModelTime() {
...
-              .getTransactions(offset, filter, false)
-              .filter(x -> !x.getEvents().isEmpty())
-              .blockingFirst();
-      return transaction.getEvents().stream()
-          .filter(x -> x instanceof CreatedEvent)
-          .map(x -> (CreatedEvent) x)
-          .collect(Collectors.toList());
-    }
-    return activeContractSetResponse.getCreatedEvents();
+  // TODO: Rework polling from HTTP and TimeUpdaterBot scheduling.
```
Comment (richardkapolnai-da) [inline src/main/java/com/digitalasset/refapps/marketdataservice/timeservice/TimeUpdaterBot.java]: let's jira this

File: src/main/java/com/digitalasset/refapps/marketdataservice/utils/CliOptions.java [hunk 11]
```diff
@@ -11,14 +11,18 @@
 import org.kohsuke.args4j.Option;
 import org.kohsuke.args4j.spi.StringArrayOptionHandler;
 
-// Valid use case for a DataClass.
-@SuppressWarnings("PMD.DataClass")
+// Valid use case for a DataClass. CLI parser requires non-fields for options.
```
Comment (richardkapolnai-da) [inline src/main/java/com/digitalasset/refapps/marketdataservice/utils/CliOptions.java]:
```suggestion
// CLI parser requires non-fields for options.
```

File: src/main/resources/logback.xml [hunk 12]
```diff
@@ -20,6 +20,7 @@
     </encoder>
   </appender>
   <logger name="com.digitalasset.refapps" level="debug" />
+  <logger name="jsonapi" level="debug" />
```
Comment (richardkapolnai-da) [inline src/main/resources/logback.xml]: remove or:
```suggestion
  <logger name="jsonapi" level="info" />
```

File: src/test/java/com/digitalasset/jsonapi/JsonLedgerClientIT.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+  public void setUp() {
+    ledger = sandbox.getLedgerAdapter();
+    ledgerId = sandbox.getClient().getLedgerId();
+    String jwt =
+        Jwt.createToken(ledgerId, APPLICATION_ID, Collections.singletonList(OPERATOR.getValue()));
+    httpClient = new ApacheHttpClient(httpResponseDeserializer, jsonSerializer, jwt);
+    webSocketClient = new TyrusWebSocketClient(webSocketResponseDeserializer, jsonSerializer, jwt);
+    api = new Api("localhost", 7575);
+  }
+
```
Comment (richardkapolnai-da) [inline src/test/java/com/digitalasset/jsonapi/JsonLedgerClientIT.java]: test `create`

File: src/test/java/com/digitalasset/jsonapi/apache/ApacheHttpClientIT.java [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    ledger.createContract(party, CurrentTime.TEMPLATE_ID, currentTime.toValue());
+    CurrentTime.ContractId contractId =
+        ledger.getCreatedContractId(party, CurrentTime.TEMPLATE_ID, ContractId::new);
+    ExerciseCommand command = contractId.exerciseCurrentTime_AddObserver("MarketDataVendor");
+
+    HttpClient client = new ApacheHttpClient(deserializer, new GsonSerializer(), jwt);
+    HttpResponse response = client.post(api.exercise(), command);
+
+    assertThat(response.getStatus(), is(200));
+  }
```
Comment (richardkapolnai-da) [inline src/test/java/com/digitalasset/jsonapi/apache/ApacheHttpClientIT.java]: unite these, and rename to sth like `post`
move to JsonLedgerClientIT

File: src/test/java/com/digitalasset/jsonapi/apache/ApacheHttpClientIT.java [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+  private String createJwt(String ledgerId, List<String> parties) {
```
Comment (richardkapolnai-da) [inline src/test/java/com/digitalasset/jsonapi/apache/ApacheHttpClientIT.java]: reuse the lib jwt thingy?

File: Dockerfile-json (hunk 1 above)
Comment (tamaskalcza-da) [reply Dockerfile-json]: https://github.com/DACH-NY/refapp-maintenance/pull/74

File: src/main/java/com/digitalasset/jsonapi/gson/GsonDeserializer.java (hunk 6 above)
Comment (tamaskalcza-da) [reply src/main/java/com/digitalasset/jsonapi/gson/GsonDeserializer.java]: https://digitalasset.atlassian.net/browse/ERA-745

File: src/main/java/com/digitalasset/refapps/marketdataservice/timeservice/TimeUpdaterBot.java (hunk 10 above)
Comment (tamaskalcza-da) [reply src/main/java/com/digitalasset/refapps/marketdataservice/timeservice/TimeUpdaterBot.java]: https://digitalasset.atlassian.net/browse/ERA-746

**Pull request merged**